### DISCUSSION TITLE
About Moose Handling Heat equation in two domain?

### DISCUSSION THREAD

**Discussion by ashokgrg19** [idaholab/moose#23823] [Q&A Modules: General] [answered]
I am trying to model a solidification process in MOOSE. 
This is how I approached the problem:
1. Created the two domains using mesh generator: The inside rectangle is the casting( primary block) b) The outside rectangle is the mold (secondary block). Please have a look into the attached input file. 
2. Solved the heat equation in respective blocks: 
                      For Casting
                       dH/dT_c*dT_c/dt = \nabla. k_c \nabla T_c ; T_c, k_c and H stand for temperature, thermal conductivity and Enthalpy of  casting 
                      For Mold
                      rho_m*C_m*dT_m/dt=\nabla. k_m \nabla T_m; T_m, k_m, rho_m and C_m stand for temperature, thermal conductivity , density and specific heat capacity of mold 
     with my interface conditions:
     T_m = T_c ..................................................................i
      and
       -k_m dT_m/dn = -k_c dT_c/dn..................................ii
 and Dirichlet boundary condition at mold surface    
 
3. I used MatchedValueBC for condition (i) and   InterfaceDiffusion kernel  for condition (ii)

4. On doing so, temperature of the casting T_c remains fixed while T_m evolves with the time. 
 I am not sure why this is happening. Actually, T_c should be decreasing while T_m should be increasing with the time. Not sure what I am missing here. 
 
 Could you please have a look into my input file and let me know if I am missing anything here? 
 
 Would really appreciate any help and suggestions!
 Thank you so much for your time! 
 ```
 [Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 100
    xmax = 0.6
    ny = 100
    ymax = 0.6
  []
  [subdomain1]
    input = gen
    type = SubdomainBoundingBoxGenerator
    bottom_left = '0.2 0.20 0'
    top_right = '0.40 0.40 0'
    block_id = 1
  []
  [interface]
    type = SideSetsBetweenSubdomainsGenerator
    input = subdomain1
    primary_block = '1'
    paired_block = '0'
    new_boundary = 'primary1_interface'
  []
[]

[Variables]
  [u] #casting
    initial_condition = 500
    block = 1
  []
  [v] #Mold
    initial_condition = 17
    block = 0
  []
[]

[Kernels]
  [./heat_conductionAlloy]
    type = HeatConduction
    variable = u
    block=1
  [../]
  [./heat_conductionMold]
    type = HeatConduction
    variable = v
    block=0
  [../]
  [./heat_conduction_time_derivative]
    type = HeatConductionTimeDerivative
    variable = u
    block=1
  [../]
  [./heat_conduction_time_derivativeMold]
    type = HeatConductionTimeDerivative
    variable = v
    block=0
  [../]
[]

[BCs]
  [u]
    type = DirichletBC
    variable = v
    boundary = 'left bottom right top'
    value= 17
  []
  [interface_bc]
    type = MatchedValueBC
    variable = v
    v = u
    boundary = primary1_interface
  []
[]

[Preconditioning]
  [./smp]
    type = SMP
    full = true
  [../]
[]
[Problem]
  type = FEProblem
[]
[Executioner]
  type = Transient
    [./TimeStepper]
    type = IterationAdaptiveDT
    dt = 1
    #iteration_window = 3
    optimal_iterations = 5 #should be 5 probably
    growth_factor = 1.4
    linear_iteration_ratio = 1000
    cutback_factor =  0.8
  [../]
  num_steps = 4000
  dtmax=10
  solve_type = 'PJFNK'       #"PJFNK, JFNK, NEWTON"
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
  scheme = 'crank-nicolson'   #"implicit-euler, explicit-euler, crank-nicolson, bdf2, rk-2"
[]

[Outputs]
  exodus = true
  print_linear_residuals = true
[]

[InterfaceKernels]
  active = 'diffusion'
  [./diffusion]
    type = InterfaceDiffusion
    variable = v
    neighbor_var = u
    boundary = primary1_interface
    D = 'D'
    D_neighbor = 'D'
  [../]
[]

[Materials]
  [mat0]
    type = GenericConstantMaterial
    prop_names = 'D'
    prop_values = '-0.2'
    block = 0
  []
  [mat1]
    type = GenericConstantMaterial
    prop_names = 'D'
    prop_values = '-211'
    block = 1
  []
  [./Alloy]
    type = GenericConstantMaterial
    prop_names = 'thermal_conductivity specific_heat density'
    prop_values = '211 1190 2555'
    block = 1
  [../]
  [./Mold]
    type = GenericConstantMaterial
    prop_names = 'thermal_conductivity specific_heat density'
    prop_values = '0.2 1000 1380'
    block = 0
  [../]
[]
[Postprocessors]
  [./0.00_temperature]
    type = PointValue
    point = '0.22 0.22 0.0'
    variable = T
  [../]
  [./03_temperature]
    type = PointValue
    point = '0.18 0.18 0.0'
    variable = T
  [../]
[]

[AuxVariables]
  [T][]
[]

[AuxKernels]
  [uK]
    type = ParsedAux
    variable = T
    args = 'u'
    function = 'u'
    block = 1
  []
  [vK]
    type = ParsedAux
    variable = T
    args = 'v'
    function = 'v'
    block = 0
  []
[]

 ```

**Comment by GiudGiud** [comment]
I m wondering if you need to duplicate the MatchedValueBC so that it comes in the equation for u as well
```
 [interface_bc]
   type = MatchedValueBC
   variable = v
   v = u
   boundary = primary1_interface
 []
 [interface_bc]
   type = MatchedValueBC
   variable = u
   v = v
   boundary = primary1_interface
 []
```

just an idea

**Reply by GiudGiud** [reply]
is there supposed to be a source term or should this be all driven by the external dirichletBCs ?

**Reply by lindsayad** [reply]
One `MatchedValueBC` will enforce the condition that u = v. You don't need to enforce it twice

**Reply by GiudGiud** [reply]
alright so it s not that

**Reply by GiudGiud** [reply]
A matched value when the initial conditions are so different seems odd.
Can you try with just a diffusion term?

Besides, if you enforce continuity and a diffusion flux, maybe you should just use a single variable

**Reply by lindsayad** [reply]
> Besides, if you enforce continuity and a diffusion flux, maybe you should just use a single variable

Excellent suggestion.

I can tell you that if you swap `u` and `v` in your `InterfaceDiffusion` `variable` and `neighbor_var`, then stuff starts happening. `variable` should always correspond to the variable defined on the subdomain associated with the sideset you pass to `boundary`. I thought we had checks to help detect a bad setup like you currently have but I guess not.

**Reply by lindsayad** [reply]
> A matched value when the initial conditions are so different seems odd.

Yes. This actually creates a boundary shock in your `v` profile

![Screenshot from 2023-03-23 16-19-45](https://user-images.githubusercontent.com/10248304/227385583-c85342d6-ba68-4ed2-8acd-247df213b544.png)

**Reply by lindsayad** [reply, marked as answer]
So I made the following changes:
1. Changed to use `PenaltyInterfaceDiffusion` which removes the need for the `MatchedValueBC`
2. Changed from `crank-nicolson` to `bdf2`. The latter seems to be much less prone to both spatial and temporal oscillations
```
diff --git a/modules/heat_conduction/user.i b/modules/heat_conduction/user.i
index 03d37fa427..dfebdc6875 100644
--- a/modules/heat_conduction/user.i
+++ b/modules/heat_conduction/user.i
@@ -64,12 +64,12 @@
    boundary = 'left bottom right top'
    value= 17
  []
- [interface_bc]
-   type = MatchedValueBC
-   variable = v
-   v = u
-   boundary = primary1_interface
- []
+ # [interface_bc]
+ #   type = MatchedValueBC
+ #   variable = v
+ #   v = u
+ #   boundary = primary1_interface
+ # []
 []
 
 [Preconditioning]
@@ -97,7 +97,7 @@
  solve_type = 'PJFNK'       #"PJFNK, JFNK, NEWTON"
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
- scheme = 'crank-nicolson'   #"implicit-euler, explicit-euler, crank-nicolson, bdf2, rk-2"
+ scheme = 'bdf2'   #"implicit-euler, explicit-euler, crank-nicolson, bdf2, rk-2"
 []
 
 [Outputs]
@@ -108,12 +108,11 @@
 [InterfaceKernels]
  active = 'diffusion'
  [./diffusion]
-   type = InterfaceDiffusion
-   variable = v
-   neighbor_var = u
+   type = PenaltyInterfaceDiffusion
+   variable = u
+   neighbor_var = v
    boundary = primary1_interface
-   D = 'D'
-   D_neighbor = 'D'
+   penalty = 1e6
  [../]
 []
 
```
and things look pretty good after 80 time steps

![Screenshot from 2023-03-23 16-52-08](https://user-images.githubusercontent.com/10248304/227389294-e205445c-0e31-49fa-a8f9-a1c3b64643d1.png)

**Reply by ashokgrg19** [reply]
Thanks @lindsayad and @GiudGiud for helpful suggestions. It worked on normal heat conduction where there is no phase change involved (liquid to solid).

 Yes, the initial condition looks bit weird. However, the initial temperature of the mold and cast largely differ on their scale, former one is at room temperature and later one is very hot--liquid phase. 

I have one question regarding the solution suggested: What's the difference between the InterfaceDiffusion and PenaltyInterfaceDiffusion? Looking on moose documentation PenaltyIntefraceDiffusion handles the continuity between the domains providing penalty value, but other does not. So, could you please explain me which condition these two kernels are suitable?

Thank you again for the help and suggestion.   
Ashok

**Reply by GiudGiud** [reply]
The differences are tied to the numerical scheme you want to use. 

the intent is similar between the two methods, though InterfaceDiffusion only ensures a diffusive flux, while the second also attempts to impose variable continuity.

**Reply by ashokgrg19** [reply]
Thank you so much @GiudGiud for explaining this!
